source/includes/read/retrieve.c

```diff
@@ -0,0 +1,72 @@
+#include <bson/bson.h>
+#include <mongoc/mongoc.h>
+#include <stdio.h>
+
+int
+main (int argc, char *argv[])
+{
+    mongoc_client_t *client;
+    mongoc_collection_t *collection;
+    bson_error_t error;
+
+    mongoc_init ();
+
+    client =
+        mongoc_client_new ("<connection string URI>");
+    collection = mongoc_client_get_collection (client, "sample_restaurants", "restaurants");
+
+    {
+        // start-find
+        bson_t *filter = BCON_NEW ("cuisine", BCON_UTF8 ("Spanish"));
+
+        mongoc_cursor_t *results = 
+            mongoc_collection_find_with_opts (collection, filter, NULL, NULL);
+        // end-find
+        
+        // start-find-iterate
+        char *str;
+        const bson_t *doc;
+
+        while (mongoc_cursor_next (results, &doc)) {
+            str = bson_as_canonical_extended_json (doc, NULL);
+            printf ("%s\n", str);
+            bson_free (str);
+        }
```

Comment:
Consider showing error checking for a cursor. If `mongoc_cursor_next` returns false, that can be either because there are no documents left, or an error occurred:

```suggestion
        }

        bson_error_t error;
        if (mongoc_cursor_error(results, &error))
        {
            fprintf(stderr, "Error getting results: %s\n", error.message);
        }
```